src/js/components/conversation.jsx

```diff
@@ -62,7 +62,32 @@ export class ConversationComponent extends Component {
     }
 
     render() {
-        const messages = this.props.conversation.messages.map((message) => {
+        const processedMessages = [...this.props.conversation.messages];
+        let lastAuthor;
+        processedMessages.forEach((message, index) => {
+            const author = message.role === 'appUser' ? message.role : `${message.role}_${message.name}`;
+
+            if (!lastAuthor) {
+                lastAuthor = author;
+                message.firstInGroup = true;
```

Comment:
Yes you could do that, but then you need to reprocess the message array on each render. What I'm suggesting is to do that processing in the reducer when doing `ADD_MESSAGE` or `SET_CONVERSATION`. That way the processing happens only when it's really needed. The component can rerender when you don't really expect it (change of state somewhere, change a flag, etc.). 

And yes, it can belong in the reducer... you can pretty much put anything in the reducer. The reducer is also a good place to put state instead of using local component state.